### ISSUE TITLE
Feature Request - add a `build_hooks` rez-config option

### ISSUE THREAD

**Issue by ColinKennedy** [nerdvegas/rez#1239]
Rez currently has a configurable [release_hooks](https://github.com/nerdvegas/rez/wiki/Configuring-Rez#release_hooks). A Rez release hook can [run pre_release](https://github.com/nerdvegas/rez/blob/6bd2f451e2ca464abcdcca3c7060a81fca1c2765/src/rezplugins/build_process/local.py#L87-L93), before any packages are built or any requirement packages are resolved or [on post_release](https://github.com/nerdvegas/rez/blob/6bd2f451e2ca464abcdcca3c7060a81fca1c2765/src/rezplugins/build_process/local.py#L106-L112), to do some kind of post release action.

So in effect, `ReleaseHookEvent.pre_release` functions exactly like a pre-build hook. Except it only runs during `LocalBuildProcess.release`. I would like to see the same mechanism added but instead added to `LocalBuildProcess.build`.

## The use-case for why I think this is a good addition

I want to query and configure a Rez package just before build. Normally this is solvable by adding an `early()` or `late()` call to the package. However in my case, I'm finding that 

A. I am adding the same early function to all of my Rez packages and 
B. That function doesn't vary at all between packages. 
C. I always want this early() function whenever the package has a `"publish_documentation"` rez-test key.

Even that might be alright, as long as the `early()` function is set once and then forgotten. However in the future the `early()` function has a strong chance of needing to be changed, which means I'd have to update dozens or even hundreds of packages with the new edit - everywhere that has that `early()` call.

Considering that the content does not vary and it is always needed whenever the `"publish_documentation"` rez-test key is present, that seems like the perfect opportunity for a pre-build hook. That way, the hook can be applied across packages either globally or using `with scope("config") as config:` and when the plugin needs to change in the future, none of the Rez package need to be updated.

What do you think of this use case? Is it solvable without modifying `LocalBuildProcess.build` or do you think this is a worthy addition? IMO it's worthy because `LocalBuildProcess.release` already has what I'm looking for, via `release_hooks`. We just don't have the equivalent code in `LocalBuildProcess.build` yet.

**Comment by ColinKennedy** [comment]
Hey Fede,

Honestly I wasn't aware of `preprocess`. I kept searching for "pre_" in the wiki files so it didn't come up :P. 

I'll need to try it myself to know for sure but I do have some concerns, judging from the documentation...

- Will I be able to import and load external Python packages in the body of that preprocess function? I know Rez has some tight controls over what is and is not importable at times. Or maybe this is a case where `@include("my_python_namespace")` will work.
- What if I want more than one global preprocess function? I'd have to wrap the functionality of both functions into one uber preprocess in that case, I think. It's doable but doesn't feel very clean IMO.
- It seems if  `package_preprocess_mode` is set to "override" (which is the default), a package-defined preprocess will omit any global preprocess. So clients of my preprocess would need to explicitly set `package_preprocess_mode` to "before" or "after"  to get both global preprocess still runs alongside the user's preprocess function. Not a big deal but does require some effort.

In the case of plug-ins, it's a cleaner solution I feel because it requires no changes at the package level and from my tests, it's more flexible with what you can import.

Anyway, those are concerns just from the documentation. They may not be problems in practice. I'll try this out sometime next week so I'll know for sure. Thank you for the link!

**Comment by ColinKennedy** [comment]
Quick update: `package_preprocess_function` is passed `this` and `data` by the caller. But unfortunately, it passes the in-memory installed Rez package, not the source package. In my case, I need...

1. The source directory from which the built / installed Rez package came from
2. A mechanism to edit the built / installed package, using info that I query from the source directory.

`preprocess` / `package_preprocess_function` satisfies 2 but from what I can see I cannot get 1 from the solution.

In contrast, the `ReleaseHook` has a reference to the source path in the instance itself plus its methods `pre_release` / `post_release` include `variants` + `install_path`. 

I think I can't get what I need using regular preprocess. A `pre_build` via `build_hooks` would be needed.

**Comment by nerdvegas** [comment]
Unfortunately I don't think you're going to be able to modify the package definition via release_hook, even if that is extended to support pre-build - release hooks weren't written with package modification in mind, this is what `preprocess` is for.

A good approach may be to update `package_preprocess_function` so that it's able to call a py function within a separately resolved env, something like so:

```
package_preprocess_function = {
    "request": ["my_build_utils-1"],
    "function": "my_build_utils.preprocess"
}
```

It's not that preprocessing-as-a-plugin (via modification ability in some PreBuildHook) is a bad idea, it's just that I suspect it's significantly more work, and it would also muddy the waters between what should be a pre build hook and what should be a preprocessor.

**Comment by ColinKennedy** [comment]
Alright, it looks like I either need to simplify my set-up or provide something more concrete to show and discuss. Thank you @fnaum and @nerdvegas for your suggestions, I'm going to close this and see workshop a bit more.

**Comment by ColinKennedy** [comment]
Actually one last thing about `package_preprocess_function` - it appears to be applying to my in-memory DeveloperPackage. That's a bug, right? `package_preprocess_function` and `preprocess` in the Rez Wiki explicitly mentions running as a part of the build process. But if you run this code

```python
from rez.config import config
from rez import developer_package


def _foo(*_, **__):
    another_path = "/tmp/another_package_folder"
    developer_package.DeveloperPackage.from_path(another_path)
    print("RUNNING!")


config.package_preprocess_function = _foo
path = "/tmp/some_package_folder"
package = developer_package.DeveloperPackage.from_path(path)
print(package)
```

It either errors with `'maximum recursion depth exceeded in comparison` or seg-faults with `Aborted (core dumped)` before it can print the traceback.

I can log this as a separate issue if needed.

**Comment by nerdvegas** [comment]
Hey Colin,

I'm not sure what you're doing here, why are you creating a DeveloperPackage instance within the `_foo` preprocessor func? That would be the source of the infinite recursion.

The preprocessor just takes a Package instance (this), and the raw package data (data), and allows you to change `data` in-place in order to modify the package definition before it gets installed. If there was a bug then the tests (see test_packages) would be picking it up.

**Comment by ColinKennedy** [comment]
For my specific use case, I don't explicitly need a `DeveloperPackage` within `_foo`. I just stumbled upon that recursive loop issue while trying some things out and was surprised because preprocessor in the documentation is described as only running during build. So it running during `DeveloperPackage` seemed like maybe a mistake. 

As I mentioned, I went a different way so I don't think I call `DeveloperPackage` within `_foo` anymore. But I wanted to make a note here. Maybe Rez's documentation needs updating if `preprocess` isn't just a build-time only process. Or log it as a bug in case that's not expected.
